src/common/state/table.h

```diff
@@ -12,8 +12,19 @@ typedef struct table_callback_data table_callback_data;
 typedef void *table_done_callback;
 
 /* The callback called when the database operation hasn't completed after
- * the number of retries specified for the operation. */
-typedef void (*table_fail_callback)(unique_id id, void *user_context);
+ * the number of retries specified for the operation.
+ *
+ * @param id The unique ID that identifies this callback. Examples include an
+ *        object ID or task ID.
+ * @param user_context The state context for the callback. This is equivalent
+ *        to the user_context field in table_callback_data.
+ * @param user_data A data argument for the callback. This is equivalent to the
+ *        data field in table_callback_data. The user is responsible for
+ *        freeing user_data.
+ */
+typedef void (*table_fail_callback)(unique_id id,
```

Comment:
Couldn't the signature take some sort of `void *context` and we just interpret it as an object_id or a task_id depending on the method?